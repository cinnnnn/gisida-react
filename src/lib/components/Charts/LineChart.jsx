import React from 'react';
import PropTypes from 'prop-types';
import Highcharts from 'highcharts';

class LineChart extends React.Component {
  constructor(props) {
    super(props);
    const {
      categories,
      series,
      indicator,
      pointClickCallback,
      chartTitle,
      chartHeight,
    } = this.props;

    this.state = {
      chart: {
        type: 'line',
        height: chartHeight || 200,
        backgroundColor: 'rgba(255,255,255,0)',
      },
      title: {
        text: chartTitle || null,
      },
      xAxis: {
        categories: categories.categories,
      },
      yAxis: {
        title: {
          text: indicator,
        },
        endOnTick: false,
      },
      legend: {
        enabled: false,
      },
      plotOptions: {
        dataLabels: {
          enabled: false,
        },
        series: {
          cursor: 'pointer',
          animation: false,
          events: {
            click: (e) => {
              pointClickCallback(e);
            },
          },
        },
      },
      series: series.series,
      credits: {
        enabled: false,
      },
    };
  }

  componentDidMount() {
    const self = this;
    setTimeout(() => {
      self.chart = Highcharts.chart(self.chartEl, self.state);
    }, 300);
  }

  componentWillReceiveProps(nextProps) {
    if (this.chart) {
      this.chart.destroy();
    }
    const { categories, series, indicator } = nextProps;

    this.setState({
      xAxis: {
        categories: categories.categories,
      },
      yAxis: {
        title: {
          text: indicator,
        },
      },
      series: series.series,
    }, () => {
      this.chart = Highcharts.chart(this.chartEl, this.state);
    });
  }

  componentWillUnmount() {
    if (this.chart) {
      this.chart.destroy();
    }
  }

  render() {
    return <div ref={(el) => { this.chartEl = el; }} />;
  }
}

LineChart.propTypes = {
  categories: PropTypes.objectOf(PropTypes.array).isRequired,
  series: PropTypes.objectOf(PropTypes.array).isRequired,
  indicator: PropTypes.string.isRequired,
  pointClickCallback: PropTypes.func.isRequired,
  chartTitle: PropTypes.string.isRequired,
  chartHeight: PropTypes.number.isRequired,
};

export default LineChart;
