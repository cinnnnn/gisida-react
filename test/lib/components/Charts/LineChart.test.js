import React from 'react';
import { LineChart }  from '../../../../src/lib/components/Charts/LineChart.jsx'
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';

const pointClickCallback = jest.fn();

const componentWrapper = mount(
	<LineChart 
		series={{}}
		chartTitle='ex-title'
		chartHeight={30}
		categories={{}}
		indicator=''
		pointClickCallback={pointClickCallback}
	/>
);

describe('LineChart', () => {
	it('LineChart component renders correctly', () => {
		// componentWrapper.instance().componentWillReceiveProps()
		const json = toJson(componentWrapper)
		expect(json).toMatchSnapshot();
	})
});
