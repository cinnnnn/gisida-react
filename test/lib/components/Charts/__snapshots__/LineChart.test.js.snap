// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LineChart LineChart component renders correctly 1`] = `
<LineChart
  categories={Object {}}
  chartHeight={30}
  chartTitle="ex-title"
  indicator=""
  pointClickCallback={[MockFunction]}
  series={Object {}}
>
  <div />
</LineChart>
`;
